Lean4Axiomatic/Rational/Reciprocation.lean

```diff
@@ -183,6 +182,58 @@ variable {ℚ : Type}
   [Core (ℤ := ℤ) ℚ] [Addition ℚ] [Multiplication ℚ]
   [Negation ℚ] [Reciprocation ℚ] [Division ℚ]
 
+/--
+Equivalent to `Division.Props.ind_fraction` but with a more convenient argument
+order when using the `apply` tactic.
+-/
+def ind_fraction_on
+  {motive : ℚ → Prop}
+  [AA.Substitutive₁ (α := ℚ) (motive) (· ≃ ·) (· → ·)] (p : ℚ) 
+    (on_int_frac : ({a b : ℤ} → [Integer.Nonzero b] → motive ( (a : ℚ) / b)))
+    : motive p :=
+    ind_fraction on_int_frac p
+
+/--
+The predicate AsRatio satisfies a substitivie property with respect to the
+equivalence relation ≃. 
+I.e. If two rationals a and b are equvalent and a can be expressed as an
+integer ratio, a ≃ a' / b', then b can be expressed in the same way.
+-/
+theorem AsRatio_prop_subst_lemma : ∀ {a b : ℚ}, a ≃ b → AsRatio a → AsRatio b
+    := by
+  intro a b aeb as_ratio_a 
+  exact match as_ratio_a with
+  | AsRatio.intro a' b' bnz eqfrac =>
+    have b_as_ratio : b ≃ a' / b' := calc
+      b      ≃ _ := eqv_symm aeb
+      a      ≃ _ := eqfrac
+      from_integer a' / from_integer b'  ≃ _ := eqv_refl
```

Comment:
There are a few tweaks to make here:
1. No need to name the result (`b_as_ratio`), because you can refer to it below using `‹b ≃ a' / b'›`.
2. References to argument names (`aeb`, `eqfrac`) should be replaced with types (`‹a ≃ b›`, `‹a ≃ a' / b'›`).
3. Switch to the style of putting the "underscore side" of the relations first (see below); this is more readable because the proofs then follow the calculations they justify.
4. Align all of the proofs on the same column (easier to read).
5. Only use casting notation or conversion functions (`from_integer`) when absolutely necessary, or if there's a lot of casting going on and being explicit makes things clearer. I think in this case `a' / b'` shouldn't cause any problems.

Putting all of those together:

```lean4
  have : b ≃ a' / b' := calc
    _ ≃ b       := eqv_refl
    _ ≃ a       := eqv_symm ‹a ≃ b›
    _ ≃ a' / b' := ‹a ≃ a' / b'›
```